### PR TITLE
balena-cli: 16.7.6 -> 17.0.0; Build from source

### PR DESCRIPTION
**Pull request by doronbehar**
###### Motivation for this change

Tested `balena login`, and `balena scan`, though without devices attached. All seems to be functional.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Result of `nixpkgs-review pr 252281` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>balena-cli</li>
  </ul>
</details>

**Comment by doronbehar** [comment]
Thanks for the review @matthiasbeyer. Will merge in a few days if no respond will come from maintainer @kalebpace .

**Comment by kalebpace** [comment]
Unfortunately, some of the dependencies needed to build this are not supported on `aarch64-darwin` or `x86_64-darwin`
 
This work should definitely be included though. Could you add a conditional for the platform and instead install the binary for the platforms mentioned above? 

Thanks for your time and work on this!

image

**Comment by doronbehar** [comment]
@kalebpace could you try it now? I added the `udev` dependency on the condition `!stdenv.isDarwin`

**Comment by kalebpace** [comment]
> @kalebpace could you try it now? I added the `udev` dependency on the condition `!stdenv.isDarwin`

Results of `nix build .#balena-cli`
image

**Comment by doronbehar** [comment]
Thanks for the quick report! Pushed another attempt.

**Comment by doronbehar** [comment]
I'm fairly optimistic we can (together) also get the Darwin build from source to succeed, which will enable us not to maintain another extra hash and a much more complicated expression.

**Comment by doronbehar** [comment]
@ofborg build balena-cli

**Comment by kalebpace** [comment]
Same to you! Will help as quickly as I can. 

Attempted another build, similar but different error, as it partially completed installing dependancies

head
image

tail
image

**Comment by doronbehar** [comment]
Hmmm... First of all, you may find it easier to use, after a build fails:

```
nix log .#balena-cli | nix run nixpkgs#ansifilter | pbcopy
```

It might require enabling `experimental-features = nix-command flakes` in `~/.config/nix/nix.conf`.

And as for the error, could you check if `nix build -Lf. darwin.cctools` contains a `libtool` executable in `result/bin/`? I tried to put it in `nativeBuildInputs` now...

@ofborg build balena-cli

^ In case you won't be quick this time :)

**Comment by doronbehar** [comment]
We've reached some progress!

```
../src/darwin/REDiskList.h:21:9: fatal error: 'Foundation/Foundation.h' file not found
npm ERR! #import <Foundation/Foundation.h>
npm ERR!         ^~~~~~~~~~~~~~~~~~~~~~~~~
npm ERR! 1 error generated.
```

I pushed another fix attempt. Perhaps more Darwin specific `buildInputs` will be needed now. Naturally it will be much easier with 1st hand access to a Darwin machine. Until then:

@ofborg build balena-cli

**Comment by doronbehar** [comment]
@ofborg build balena-cli

**Comment by doronbehar** [comment]
Nice, CI is fast today...

@ofborg build balena-cli

**Comment by doronbehar** [comment]
@kalebpace this seems ready for Darwin now. Your testing and approval will be appreciated.

File: pkgs/tools/admin/balena-cli/default.nix
Comment (wegank) [inline pkgs/tools/admin/balena-cli/default.nix:38]:
```suggestion
```

Darwin binaries are in Mach-O format, not ELF, so `patchelf` should have no effect.

**Comment by kalebpace** [comment]
> ^ In case you won't be quick this time :)

My apologies, dozed off shortly after I responded last night 😬 Glad CI was fast/available! 

Builds and prints version for me, thanks again for seeing this through. Should be good to go!

**Review by kalebpace — APPROVED** [review]

File: pkgs/tools/admin/balena-cli/default.nix
Comment (doronbehar) [reply pkgs/tools/admin/balena-cli/default.nix]: > Darwin binaries are in Mach-O format, not ELF, so `patchelf` should have no effect...?

I wasn't sure, thanks for your help. Removed it.

**Review by kalebpace — APPROVED** [review]

**Pull request merged**
